File: app/XDM/XDM.App.Core/HttpServer/RequestContext.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Text;
#if !NET5_0_OR_GREATER
using NetFX.Polyfill;
#endif

namespace HttpServer
{
    public class RequestContext
    {
        private TcpClient tcp;
        public string RequestPath { get; }
        public byte[]? RequestBody { get; }
        public Dictionary<string, List<string>> RequestHeaders { get; }
        public byte[]? ResponseBody { set; get; }
        public Dictionary<string, List<string>> ResponseHeaders { set; get; }
        public ResponseStatus ResponseStatus { set; get; }
        public bool KeepAlive { get; private set; }

        internal RequestContext(string path, Dictionary<string, List<string>> headers, byte[]? body, TcpClient tcp, bool keepAlive)
        {
            this.RequestPath = path;
            this.RequestHeaders = headers;
            this.RequestBody = body;
            this.tcp = tcp;
            this.ResponseHeaders = new();
            this.ResponseStatus = new ResponseStatus { StatusCode = 200, StatusMessage = "OK" };
            this.KeepAlive = keepAlive;
        }

        public void SendResponse()
        {
            var io = this.tcp.GetStream();
            var responseBuffer = new StringBuilder();
            responseBuffer.Append($"HTTP/1.0 {this.ResponseStatus.StatusCode} {this.ResponseStatus.StatusMessage}\r\n");
            foreach (var headerName in ResponseHeaders.Keys)
            {
                if (headerName.Equals("content-length", StringComparison.InvariantCultureIgnoreCase))
                {
                    continue;
                }
                foreach (var value in ResponseHeaders[headerName])
                {
                    responseBuffer.Append($"{headerName}: {value}\r\n");
                }
            }
            responseBuffer.Append($"Connection: keep-alive\r\n");
            if (ResponseBody != null && ResponseBody.Length > 0)
            {
                responseBuffer.Append($"Content-Length: {ResponseBody.Length}\r\n");
            }
            responseBuffer.Append("\r\n");
            var bytes = Encoding.UTF8.GetBytes(responseBuffer.ToString());
            io.Write(bytes, 0, bytes.Length);
            if (ResponseBody != null && ResponseBody.Length > 0)
            {
                io.Write(ResponseBody, 0, ResponseBody.Length);
            }
            io.Flush();
        }

        public void AddResponseHeader(string name, string value)
        {
            var values = this.ResponseHeaders.GetValueOrDefault(name, new List<string>(1));
            values.Add(value);
            this.ResponseHeaders[name] = values;
        }
    }
}

File: xdm-app/src/main/java/xdm/integration/RequestContext.kt
package xdm.integration


import java.net.Socket
import java.nio.charset.StandardCharsets

const val CRLF = "\r\n"

class RequestContext(
    val requestMethod: String,
    val requestPath: String,
    val requestHeaders: Map<String, List<String>>,
    val requestBody: ByteArray?,
    private val socket: Socket,
    keepAlive: Boolean
) {

    /**
     * Whether the socket stays open for another request. Initialised from the request's HTTP
     * version and `Connection` header, but the handler may force it to `false` to make the response
     * the last one on this connection (and so end the thread serving it).
     */
    var keepAlive: Boolean = keepAlive

    var responseBody: ByteArray? = null

    /** Insertion-ordered, so headers go out in the order they were added. */
    private val responseHeaders: MutableMap<String, MutableList<String>> = LinkedHashMap()

    var statusCode: Int = 200

    var statusMessage = "OK"

    /**
     * Writes an HTTP/1.1 response: one `Name: value` line per header value, then `Content-Length`
     * (always, 0 without a body) and a `Connection` header that matches whether the server keeps the
     * socket open for another request ([keepAlive]).
     */
    fun sendResponse() {
        val lines = mutableListOf("HTTP/1.1 $statusCode $statusMessage")
        responseHeaders
            .filterKeys { !it.equals("Content-Length", ignoreCase = true) && !it.equals("Connection", ignoreCase = true) }
            .forEach { (name, values) -> values.forEach { lines += "$name: $it" } }
        lines += "Content-Length: ${responseBody?.size ?: 0}"
        lines += "Connection: ${if (keepAlive) "keep-alive" else "close"}"

        val io = socket.getOutputStream()
        io.write((lines.joinToString(CRLF) + CRLF + CRLF).toByteArray(StandardCharsets.UTF_8))
        responseBody?.let { io.write(it, 0, it.size) }
        io.flush()
    }

    fun getRequestHeader(name: String): String? =
        requestHeaders.entries.firstOrNull { it.key.equals(name, ignoreCase = true) }?.value?.firstOrNull()

    fun addResponseHeader(name: String, value: String) {
        val values = responseHeaders.getOrDefault(name, ArrayList())
        values.add(value)
        responseHeaders[name] = values
    }
}
